flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableInputFormat.java

```diff
@@ -0,0 +1,275 @@
+			jobConf, new HadoopDummyReporter());
+		if (this.recordReader instanceof Configurable) {
+			((Configurable) this.recordReader).setConf(jobConf);
+		}
+		key = this.recordReader.createKey();
+		value = this.recordReader.createValue();
+		this.fetched = false;
+		try {
+			deserializer = (Deserializer) Class.forName(sd.getSerdeInfo().getSerializationLib()).newInstance();
+			Configuration conf = new Configuration();
+			//properties are used to initialize hive Deserializer properly.
+			Properties properties = HiveTableUtil.createPropertiesFromStorageDescriptor(sd);
+			SerDeUtils.initializeSerDe(deserializer, conf, properties, null);
+			structObjectInspector = (StructObjectInspector) deserializer.getObjectInspector();
+			structFields = structObjectInspector.getAllStructFieldRefs();
+		} catch (Exception e) {
+			throw new FlinkHiveException("Error happens when deserialize from storage file.", e);
+		}
+	}
+
+	@Override
+	public HiveTableInputSplit[] createInputSplits(int minNumSplits)
+			throws IOException {
+		List<HiveTableInputSplit> hiveSplits = new ArrayList<>();
+		int splitNum = 0;
+		for (HiveTablePartition partition : partitions) {
+			StorageDescriptor sd = partition.getStorageDescriptor();
+			InputFormat format;
+			try {
+				format = (InputFormat)
+					Class.forName(sd.getInputFormat(), true, Thread.currentThread().getContextClassLoader()).newInstance();
+			} catch (Exception e) {
+				throw new FlinkHiveException("Unable to instantiate the hadoop input format", e);
+			}
+			ReflectionUtils.setConf(format, jobConf);
+			jobConf.set(INPUT_DIR, sd.getLocation());
+			//TODO: we should consider how to calculate the splits according to minNumSplits in the future.
+			org.apache.hadoop.mapred.InputSplit[] splitArray = format.getSplits(jobConf, minNumSplits);
+			for (int i = 0; i < splitArray.length; i++) {
+				hiveSplits.add(new HiveTableInputSplit(splitNum++, splitArray[i], jobConf, partition));
+			}
+		}
+
+		return hiveSplits.toArray(new HiveTableInputSplit[hiveSplits.size()]);
+	}
+
+	@Override
+	public void configure(org.apache.flink.configuration.Configuration parameters) {
+
+	}
+
+	@Override
+	public BaseStatistics getStatistics(BaseStatistics cachedStats) throws IOException {
+		// no statistics available
+		return null;
+	}
+
+	@Override
+	public InputSplitAssigner getInputSplitAssigner(HiveTableInputSplit[] inputSplits) {
+		return new LocatableInputSplitAssigner(inputSplits);
+	}
+
+	@Override
+	public boolean reachedEnd() throws IOException {
+		if (!fetched) {
+			fetchNext();
+		}
+		return !hasNext;
+	}
+
+	@Override
+	public void close() throws IOException {
+		if (this.recordReader != null) {
+			this.recordReader.close();
+			this.recordReader = null;
+		}
+	}
+
+	protected void fetchNext() throws IOException {
+		hasNext = this.recordReader.next(key, value);
+		fetched = true;
+	}
+
+	@Override
+	public Row nextRecord(Row ignore) throws IOException {
+		if (!this.fetched) {
```

Comment:
can we just call `reachedEnd()` to check whether we have any more record to offer?

Reply:
Yep.